### DISCUSSION TITLE
使用Steam++解决github的 443：Timed out问题

### DISCUSSION THREAD

**Discussion by rmbadmin** [BeyondDimension/SteamTools#234] [Ideas]
## 解决目前国内在github提交代码拉取代码经常遇到 443：Timed out 超时问题。
> 相关问题:  #143  #46 #118 
下载steam++ 2.2以上版本
勾选社区加速中的github网站加速项目，选择hosts代理模式开启加速。
如果未安装证书需要照提示安装。  

![QQ截图20210513013534](https://user-images.githubusercontent.com/12234359/118019439-92d21680-b38b-11eb-9327-d171ba8c0f19.png)

开启加速后，还需要让steam++生成的证书被git信任
使用时需要让git信任windows的根证书
git 2.1.4以上版本可以使用以下设置
```
git config --global http.sslBackend schannel
```
非Windows平台可以尝试使用
```
git config --global http.sslCAInfo /etc/ssl/certs/SteamTools.Certificate.cer
```

按照上面的方式操作后，每次只要在git操作前开启Steam++加速就可以正常使用了。

**Comment by chtholly255** [comment]
Windows10按照上面的方式操作后任报错`fatal: Unsupported SSL backend 'schannel'. Supported SSL backends: openssl`

**Reply by rmbadmin** [reply]
可以尝试再openssl内添加信任，
实在不方便的话也可以考虑禁用 ssl 验证 `git config --global http.sslVerify false` 不过不太推荐这种做法

**Comment by letterk** [comment]
不建议用steam++加速github, 你会在git上遇到各种莫名其妙的问题, 尤其是ssl验证错误, 按照上面的方法设置了schannel, 没有任何效果, 直接使用代理访问github吧

**Reply by MemoryShadow** [reply]
只是需要保证一致性, 即在启用Steam++时clone仓库即可解决信任问题, 还可以在本地构建一个`~/.ssh/config`条目使用独特开头进行clone避免弄混:

```
Host *
  PubkeyAcceptedKeyTypes +ssh-rsa
  HostKeyAlgorithms +ssh-rsa
  ServerAliveInterval 60
  ForwardAgent yes
Host steampp.github.com
  HostName github.com
  User git
  IdentityFile /home/memoryshadow/.ssh/pem/github.pem
```

将IdentityFile字段替换为你的私钥文件路径, 后续即可使用对应Host进行clone, 例如:

```bash
git clone steampp.github.com:BeyondDimension/SteamTools.git
```

使用此法可以尽可能避开是否启动Steam++时的问题. 但`~/.ssh/known_hosts`中的记录在变更连接方式时仍旧需要删除

**Comment by SwaggyZhang** [comment]
> 只是需要保证一致性, 即在启用Steam++时clone仓库即可解决信任问题, 还可以在本地构建一个`~/.ssh/config`条目使用独特开头进行clone避免弄混:
> 
> ```
> Host *
>   PubkeyAcceptedKeyTypes +ssh-rsa
>   HostKeyAlgorithms +ssh-rsa
>   ServerAliveInterval 60
>   ForwardAgent yes
> Host steampp.github.com
>   HostName github.com
>   User git
>   IdentityFile /home/memoryshadow/.ssh/pem/github.pem
> ```
> 
> 将IdentityFile字段替换为你的证书路径, 后续即可使用对应Host进行clone, 例如:
> 
> ```shell
> git clone steampp.github.com:BeyondDimension/SteamTools.git
> ```
> 
> 使用此法可以尽可能避开是否启动Steam++时的问题. 但`~/.ssh/known_hosts`中的记录在变更连接方式时仍旧需要删除



> ## 解决目前国内在github提交代码拉取代码经常遇到 443：Timed out 超时问题。
> 
> > 相关问题:  #143  #46 #118
> > 下载steam++ 2.2以上版本
> > 勾选社区加速中的github网站加速项目，选择hosts代理模式开启加速。
> > 如果未安装证书需要照提示安装。
> 
> ![QQ截图20210513013534](https://user-images.githubusercontent.com/12234359/118019439-92d21680-b38b-11eb-9327-d171ba8c0f19.png)
> 
> 开启加速后，还需要让steam++生成的证书被git信任 使用时需要让git信任windows的根证书 git 2.1.4以上版本可以使用以下设置
> 
> ```
> git config --global http.sslBackend schannel
> ```
> 
> 非Windows平台可以尝试使用
> 
> ```
> git config --global http.sslCAInfo /etc/ssl/certs/SteamTools.Certificate.cer
> ```
> 
> 按照上面的方式操作后，每次只要在git操作前开启Steam++加速就可以正常使用了。

使用下列命令之后，git clone的时候还是无法读取远程仓库，请问大佬该怎么办？
```
git config --global http.sslCAInfo /etc/ssl/certs/SteamTools.Certificate.cer
```

显示：
```
git@github.com: Permission denied (publickey).
fatal: 无法读取远程仓库。

请确认您有正确的访问权限并且仓库存在。
```

**Reply by MemoryShadow** [reply]
你需要使用你在`~/.ssh/config`中设定的名称替换仓库目录下`.git/config`文件中的`git@github.com`.

例如在[示例1](https://github.com/BeyondDimension/SteamTools/discussions/234#discussioncomment-7086851)中, 我们创建了别名`steampp.github.com`, 当clone地址为`git@github.com:MemoryShadow/pyCraft.git`时, 我们就可以将URL替换为`steampp.github.com:MemoryShadow/pyCraft.git`

要注意的事情是, 使用此方式之后不需要再提供用户名了, 因为在config中进行了配置, 这看起来是一个地址, 实际上是一个别名, 如果你担心混淆, 也可以使用`steampp_github`之类的名字进行替换(别忘记修改`~/.ssh/config`).

**Reply by MemoryShadow** [reply]
很抱歉, 我在撰写回复的时候错误的描述了IdentityFile字段后应该写入的值, 这里应该是私钥文件的路径而非证书文件的路径, 我现在才发现此处的笔误, 给您带来的困扰我非常抱歉.

@SwaggyZhang

**Reply by SwaggyZhang** [reply]
好的，感谢您的慷慨帮助！

**Comment by hA0oooooo** [comment]
谢谢帮助，gpt4o应该学习一下

**Comment by elarbor** [comment]
如果能加速Google搜索就好了，哈哈哈哈哈哈哈哈

**Comment by 4Darmygeometry** [comment]
拉git时报RPC失败
gamma@gamma-pc [ ~ ] $ git clone --recurse-submodules https://github.com/4Darmygeometry/SteamTools.git
正克隆到 'SteamTools'...
错误：RPC 失败。HTTP 502 curl 22 The requested URL returned error: 502
致命错误：预期 'packfile'
